### PR TITLE
Add docs for inline io and mixins

### PR DESCRIPTION
**Pull request by JordonPhillips**
This adds documentation for inline io and mixins. This documentation was mostly pulled from the design docs, but was updated in a number of places to better fit with the rest of the smithy docs.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/source/1.0/spec/core/model.rst [hunk 1]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+are applied, from left to right. For each mixin, iterate over the mixins
+applied to the mixin in the order in which mixins are applied. When the
+evaluated shape has no mixins, the members of that shape are added to the
+resolved list of ordered members. After evaluating all the mixins of a shape,
+the members of the shape are added onto the resolved list of ordered members.
+This process continues until all mixins and the members of the starting shape
+are added to the ordered list.
+
+Given the following model:
+
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3131]: extra newline

File: docs/source/1.0/spec/core/model.rst [hunk 2]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+    service C with [B] {
+        version: "C"
+        rename: {
+            "smithy.example#OperationA": "OpA"
+            "smithy.example#OperationC": "OpC"
+        }
+        operations: [OperationC]
+    }
+
+The ``version`` property of the local shape is kept and the ``rename`` and
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3213]: Might simplify to say "The flattned equivalent of `C` with no mixins is:", and then remove the redefined operations since the other service shapes aren't defined either.

File: docs/source/1.0/spec/core/model.rst [hunk 3]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+
+Operation mixins
+----------------
+
+Operation shapes with the :ref:`mixin-trait` MAY NOT define an ``input`` or
+``output`` shape other than the :ref:`unit-type`. This is because allowing
+input and output shapes to be shared goes against the goal of the
+:ref:`input-trait` and :ref:`output-trait`.
+
+Operation shapes with the :ref:`mixin-trait` MAY define the errors shape.
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3260]:
```suggestion
Operation shapes with the :ref:`mixin-trait` MAY define errors.
```

Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3260]:
```suggestion
Operation shapes with the :ref:`mixin-trait` MAY define errors.
```

File: docs/source/1.0/spec/core/model.rst [hunk 4]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+
+.. code-block:: smithy
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3262]: Would be great to provide a slightly better example. Like "ValidatedOperation" or something that defines a ValidationException.

File: docs/source/1.0/spec/core/model.rst [hunk 5]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+    structure MixinError {}
+
+    @error("client")
+    structure MixedError {}
+
+
+Mixins in conversion tooling
+============================
+
+When converting Smithy models to OpenAPI, JSON Schema, and other models that
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:3283]: Ah, I'd reword this a bit and merge with my other suggestions around line 2992.

**Review by mtdowling — CHANGES_REQUESTED** [review]

File: docs/source/1.0/spec/core/model.rst (hunk 2 above)
Comment (JordonPhillips) [reply docs/source/1.0/spec/core/model.rst:3213]: > remove the redefined operations since the other service shapes aren't defined either

I'm not sure what you mean by this. The operations need to be there because they're still used.

File: docs/source/1.0/spec/core/model.rst [hunk 6]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+
+.. code-block:: smithy
+
+    @mixin
+    structure UserIdentifiers {
+        id: String
+    }
+
+    @mixin
+    structure AccessDetails {
```
[the remark this answers is not in the document]
Comment (JordonPhillips) [reply docs/source/1.0/spec/core/model.rst:2726]: Some mix suffxing and adjectivizing is probably appropriate, but I'm not a fan of underscoring

File: docs/source/1.0/spec/core/model.rst [hunk 7]
```diff
@@ -2692,6 +2692,602 @@ after adding a member to the ``foo`` trait:
...
+    }
+
+Mixins be any shape type, but they MUST NOT be applied to a shape of a
+different type. For example, a string mixin can be applied to a string
+shape, but not to a blob shape.
+
+.. code-block:: smithy
+
+    @mixin
+    @pattern("[a-zA-Z0-1]*")
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply docs/source/1.0/spec/core/model.rst:2781]: Did this one get resolved?

File: docs/source/1.0/spec/core/model.rst [hunk 8]
```diff
@@ -2692,6 +2692,594 @@ after adding a member to the ``foo`` trait:
...
+
+
+Mixins are an implementation detail of the model
+================================================
+
+Mixins are an implementation detail of models and are only intended to reduce
+duplication in Smithy shape definitions. Mixins do not provide any kind of
+runtime polymorphism for types generated from Smithy models. Smithy model
+transformations like code generation or converting to other model formats
+like OpenAPI SHOULD completely elide mixins by flattening the model.
```
Comment (mtdowling) [inline docs/source/1.0/spec/core/model.rst:2999]: One additional way to make this point is to call this out:
> It is a backward compatible change to remove a mixin from a shape as long as equivalent traits and members of the mixin are applied directly to the shape. Such a change should have no impact on generated artifacts like code or OpenAPI models.

**Review by mtdowling — APPROVED** [review]
